### PR TITLE
Move pruning to background thread to reduce latency

### PR DESCRIPTION
**Pull request by LukaszRozmej**
## Changes:

Currently pruning blocks block processing, which can lead to increased latency from time to time on block. This moves pruning to background thread.

This is synchronized with block processing by locking on `_dirtyNodes`, so either block can be processed or a pruning loop can spin once. This allows for thread safety to access trie nodes while not blocking block processing. This also allows reducing latency on heavy pruning with multiple spins of loop by allowing block processing to go in-between loop spins (at the cost of short-term higher memory usage?).

Note: First versions weren't using any synchronization, but this sometimes had issues and threw exceptions.

At the disposal, there is the cancellation of pruning, which waits for the current loop spin of pruning to finish and stops it from further spins.

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [ ] Yes
- [x] No

**Comments about testing , should you have some** (optional)
Test: Very big pruning cache (like 8-16GB), FullSync+pruning, maybe heavy JsonRpc load hitting the state (eth_call or tracing?), Validate we don't have regressions in snap sync

### CODE REVIEW

**Review by smartprogrammer93 — APPROVED** [review]

**Comment by smartprogrammer93** [comment]
LGTM. The locking is simple and solves any problem that could arise.

File: src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs [hunk 1]
```diff
@@ -476,17 +492,17 @@ private void PruneCurrentSet()
         /// <exception cref="InvalidOperationException"></exception>
         private void PruneCache()
         {
+
             if (_logger.IsDebug) _logger.Debug($"Pruning nodes {MemoryUsedByDirtyCache / 1.MB()}MB , last persisted block: {LastPersistedBlockNumber} current: {LatestCommittedBlockNumber}.");
             Stopwatch stopwatch = Stopwatch.StartNew();
-            List<TrieNode> toRemove = new(); // TODO: resettable
+            List<Keccak> toRemove = new(); // TODO: resettable
```
Comment (iSOcH) [inline src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs:498]: do we have an expected number of items we could pass as `capacity` here?

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs:498]: We could count nodes, but that's somewhat bad because of locks in the concurrent dictionary. Better to use class lever resettable or ArrayPoolList, but let us do that in consecutive PR.

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs]: moved to resetable list

File: src/Nethermind/Nethermind.Trie.Test/Pruning/TreeStoreTests.cs [hunk 2]
```diff
@@ -20,8 +20,8 @@ namespace Nethermind.Trie.Test.Pruning
     [TestFixture]
     public class TreeStoreTests
     {
-        private ILogManager _logManager = new OneLoggerLogManager(new NUnitLogger(LogLevel.Trace));
-
+        private readonly ILogManager _logManager = LimboLogs.Instance;
+            // new OneLoggerLogManager(new NUnitLogger(LogLevel.Trace));
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Trie.Test/Pruning/TreeStoreTests.cs:24]: Did you miss this?

File: src/Nethermind/Nethermind.Trie.Test/PruningScenariosTests.cs [hunk 3]
```diff
@@ -62,7 +62,8 @@ public class PruningContext
             [DebuggerStepThrough]
             private PruningContext(TestPruningStrategy pruningStrategy, IPersistenceStrategy persistenceStrategy)
             {
-                _logManager = new TestLogManager(LogLevel.Trace);
+                _logManager = LimboLogs.Instance;
+                    //new TestLogManager(LogLevel.Trace);
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Trie.Test/PruningScenariosTests.cs:66]: Another one

File: src/Nethermind/Nethermind.Trie.Test/TrieTests.cs [hunk 4]
```diff
@@ -29,7 +29,8 @@ public class TrieTests
         [SetUp]
         public void SetUp()
         {
-            _logManager = new NUnitLogManager(LogLevel.Trace);
+            _logManager = LimboLogs.Instance;
+             // new NUnitLogManager(LogLevel.Trace);
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Trie.Test/TrieTests.cs:33]: And another one.

**Review by asdacap — APPROVED** [review]
LGTM

File: src/Nethermind/Nethermind.Trie.Test/Pruning/TreeStoreTests.cs (hunk 2 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Trie.Test/Pruning/TreeStoreTests.cs:24]: No the tests are currently flooded with messages from those and while tests are fine they are not needed. They would only be useful when we have regression.

File: src/Nethermind/Nethermind.Trie.Test/TrieTests.cs (hunk 4 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Trie.Test/TrieTests.cs:33]: No the tests are currently flooded with messages from those and while tests are fine they are not needed. They would only be useful when we have regression.

File: src/Nethermind/Nethermind.Trie.Test/PruningScenariosTests.cs (hunk 3 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Trie.Test/PruningScenariosTests.cs:66]: No the tests are currently flooded with messages from those and while tests are fine they are not needed. They would only be useful when we have regression.

**Comment by kamilchodola** [comment]
Starting tests from this branch

**Comment by kamilchodola** [comment]
Things to verify:
Processing queue wasn't empty added to queue a new payload: 15686051 on mainnet-teku
FullSync issue

File: src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs [hunk 5]
```diff
@@ -584,7 +601,6 @@ private void CreateCommitSet(long blockNumber)
             LatestCommittedBlockNumber = Math.Max(blockNumber, LatestCommittedBlockNumber);
             AnnounceReorgBoundaries();
             DequeueOldCommitSets();
-            Prune();
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs:587]: why removed?

File: src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs [hunk 6]
```diff
@@ -516,16 +534,13 @@ private void PruneCache()
                 }
             }
 
-            foreach (TrieNode trieNode in toRemove)
+            for (int index = 0; index < _nodesToRemoveOnPruning.Count; index++)
             {
-                if (trieNode.Keccak is null)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs:521]: why this check is no longer needed?

File: src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs (hunk 5 above)
Comment (MarekM25) [reply src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs:587]: ah I see, it was moved to FinishBlockCommit

File: src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs [hunk 7]
```diff
@@ -43,30 +44,40 @@ public DirtyNodesCache(TrieStore trieStore)
                 _trieStore = trieStore;
             }
 
-            public void SaveInCache(TrieNode node)
+            public TrieNode GetOrAdd(TrieNode node, out bool added)
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs:47]: NP: Usually the `bool` is the return value and the object is the out param.

File: src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs [hunk 8]
```diff
@@ -82,10 +93,8 @@ public TrieNode FromCachedRlpOrUnknown(Keccak hash)
                     }
 
                     // we returning a copy to avoid multithreaded access
-                    trieNode = new TrieNode(NodeType.Unknown, hash, trieNode.FullRlp);
+                    trieNode = new TrieNode(trieNode.NodeType, hash, trieNode.FullRlp);
                     trieNode.ResolveNode(_trieStore);
-                    trieNode.Keccak = hash;
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs:87]: Is this change required for the background prune? If not I recommend moving to a different PR.  Seems dangerous.

File: src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs (hunk 6 above)
Comment (kamilchodola) [reply src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs:521]: @LukaszRozmej Is this answered/addressed?

File: src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs (hunk 8 above)
Comment (kamilchodola) [reply src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs:87]: @LukaszRozmej Is this answered/addressed?

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs:87]: I am trying to figure out why https://github.com/NethermindEth/nethermind/blob/f14da669b1a70c13dabaa8dd81f9a3d56ed98644/src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs#L493-L496 is happening and this in one of my leads, that we overwrite the actual proper hash here as ResolveNode returned something different.

File: src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs (hunk 6 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs:521]: _nodesToRemoveOnPruning now contain keys only

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs:508]: Reverted
